src/main/java/bio/terra/service/dataset/StorageResource.java

```diff
@@ -1,92 +1,47 @@
+@JsonSubTypes({
+    @JsonSubTypes.Type(value = AzureStorageResource.class, name = "azure"),
+    @JsonSubTypes.Type(value = GoogleStorageResource.class, name = "gcp")
+})
+public interface StorageResource {
 
-    private UUID datasetId;
-    private CloudPlatform cloudPlatform;
-    private GoogleCloudResource cloudResource;
-    private GoogleRegion region;
+    UUID getDatasetId();
 
-    public StorageResourceModel toModel() {
-        GoogleRegion regionOverride = region;
-        if (cloudResource == GoogleCloudResource.FIRESTORE) {
-            regionOverride = region.getRegionOrFallbackFirestoreRegion();
-        } else if (cloudResource == GoogleCloudResource.BUCKET) {
-            regionOverride = region.getRegionOrFallbackBucketRegion();
-        }
-        return new StorageResourceModel()
-            .cloudPlatform(cloudPlatform)
-            .cloudResource(cloudResource.toString())
-            .region(regionOverride.toString());
-    }
-
-    public UUID getDatasetId() {
-        return datasetId;
-    }
-
-    public StorageResource datasetId(UUID datasetId) {
-        this.datasetId = datasetId;
-        return this;
-    }
-
-    public CloudPlatform getCloudPlatform() {
-        return cloudPlatform;
-    }
+    StorageResource datasetId(UUID datasetId);
 
-    public StorageResource cloudPlatform(CloudPlatform cloudPlatform) {
-        this.cloudPlatform = cloudPlatform;
-        return this;
-    }
+    CloudPlatform getCloudPlatform();
 
-    public GoogleCloudResource getCloudResource() {
-        return cloudResource;
-    }
+    CloudResource getCloudResource();
 
-    public StorageResource cloudResource(GoogleCloudResource cloudResource) {
-        this.cloudResource = cloudResource;
-        return this;
-    }
+    StorageResource cloudResource(CloudResource cloudResource);
 
-    public GoogleRegion getRegion() {
-        return region;
-    }
+    CloudRegion getRegion();
 
-    public StorageResource region(GoogleRegion region) {
-        this.region = region;
-        return this;
-    }
+    StorageResource region(CloudRegion region);
 
-    @Override
-    public boolean equals(Object o) {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
-        StorageResource that = (StorageResource) o;
-        return Objects.equals(datasetId, that.datasetId) &&
-            cloudPlatform == that.cloudPlatform &&
-            cloudResource == that.cloudResource &&
-            region == that.region;
+    default StorageResourceModel toModel() {
+        return new StorageResourceModel()
+            .cloudPlatform(getCloudPlatform())
+            .cloudResource(getCloudResource().toString())
+            .region(getRegion().toString());
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(datasetId, cloudPlatform, cloudResource, region);
+    static GoogleStorageResource getGoogleInstance() {
```

Comment:
Ah, great question.  I was doing this as a way to avoid the case where we would could create, say, a google resource and then set its cloud type to Azure.  I think given that we could just have it return a constant, I'll get rid of this static initializer